packages/core-data/src/utils/block-selection-history.ts

```diff
@@ -0,0 +1,210 @@
+/**
+ * External dependencies
+ */
+/**
+ * WordPress dependencies
+ */
+import { Y } from '@wordpress/sync';
+
+/**
+ * Internal dependencies
+ */
+import { findBlockByClientIdInDoc } from './crdt';
+import type { WPBlockSelection, WPSelection } from '../types';
+
+export const SELECTION_HISTORY_DEFAULT_SIZE = 5;
+
+export enum YSelectionType {
+	RelativeSelection = 'RelativeSelection',
+	BlockSelection = 'BlockSelection',
+}
+
+export interface YRelativeSelection {
+	type: YSelectionType.RelativeSelection;
+	attributeKey: string;
+	relativePosition: Y.RelativePosition;
+	clientId: string;
+	offset: number;
+}
+
+export interface YBlockSelection {
+	type: YSelectionType.BlockSelection;
+	clientId: string;
+}
+
+export type YSelection = YRelativeSelection | YBlockSelection;
+
+export type YFullSelection = {
+	start: YSelection;
+	end: YSelection;
+};
+
+export interface YSelectionHistory {
+	selection: YFullSelection;
+	backupSelections?: YFullSelection[];
+}
+
+/**
+ * This class is used to track recent block selections to help in restoring
+ * a user's selection after an undo or redo operation.
+ *
+ * Maintains currentSelection for the current block, and a history array for
+ * previous blocks. These properties are useful for different things:
+ *
+ * - currentSelection: The most recent selection in the current block. This is updated
+ *                     before any undo stack operations are processed.
+ * - history: An array of previous blocks. Use this as a backup for restoration
+ *            locations.
+ */
+export class BlockSelectionHistory {
+	private historySize: number;
+	private history: YFullSelection[] = [];
+	private currentSelection: YFullSelection | null = null;
+	private ydoc: Y.Doc;
+
+	constructor(
+		ydoc: Y.Doc,
+		historySize: number = SELECTION_HISTORY_DEFAULT_SIZE
+	) {
+		this.ydoc = ydoc;
+		this.historySize = historySize;
+		this.history = [];
+	}
+
+	/**
+	 * Convert a WPSelection to a Position (relative position or block position).
+	 * @param selection
+	 * @return Position or null if conversion fails
+	 */
+	private convertSelectionToPositionPart(
+		selection: WPBlockSelection
+	): YSelection {
+		const clientId = selection.clientId;
+		const block = findBlockByClientIdInDoc( clientId, this.ydoc );
+
+		const attributes = block?.get( 'attributes' ) as
+			| Y.Map< Y.Text >
+			| undefined;
```

Comment:
Assertion not necessary with `YMapWrap` :)
```suggestion
		const attributes = block?.get( 'attributes' );
```

Reply:
Merged in with your changes in https://github.com/WordPress/gutenberg/pull/73066/commits/bc818f3b8076708c8d160908f9f4cf3c0a022146.